File: 剑指offer/66.构建乘积数组.py
from typing import List


class Solution:
    def constructArr(self, a: List[int]) -> List[int]:
        pre = []
        after = []

        pre_value = 1
        for i in range(len(a)):
            if i == 0:
                pre.append(1)
            else:
                pre_value = pre_value * a[i - 1]
                pre.append(pre_value)

        after_value = 1
        for i in range(len(a)):
            if i == 0:
                after.append(1)
            else:
                after_value = after_value * a[len(a) - i]
                after.append(after_value)
        after = after[::-1]

        ans = []
        for i in range(len(a)):
            ans.append(after[i] * pre[i])
        return ans

File: 剑指offer/66.构建乘积数组.cpp
#include <vector>
using namespace std;

class Solution {
public:
    vector<int> constructArr(vector<int>& a)
    {
        vector<int> ans;
        int n = a.size();
        vector<int> num1(n, 1);
        vector<int> num2(n, 1);

        for (int i = 1; i < n; i++) {
            num1[i] = num1[i - 1] * a[i - 1];
            num2[i] = num2[i - 1] * a[n - i];
        }

        for (int i = 0; i < n; i++) {
            ans.push_back(num1[i] * num2[n - 1 - i]);
        }
        return ans;
    }
};